### ISSUE TITLE
Implement smart block (bottom) margins

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor-pdf#1515]
Block margins should only be applied between block elements. The simplest way to do this is to use top margins and only apply them if the block is not the first child. (We can come back later and implement bottom margins as an alternative).

The only tricky part are list items. List items should have a built-in line-height, so the spacing is still applied below the last item. Additionally, more space should be applied between the end of a complex block and the start of the next list item. We can try to match what we have now, though more control may be needed in the future.

One of the benefits of this change is that a block inside of another blocks (such as a block inside of an AsciiDoc table cell) will no longer have a gap below it.

**Comment by mojavelinux** [comment]
Since this change will likely break existing themes that customize block margins, it will need to be scheduled for 2.0.0.

**Comment by mojavelinux** [comment]
It may be possible to support smart margins on either side, which would mean we can avoid breaking compatibility. Regardless, it will still affect some themes since there would no longer be extra space below blocks nested inside of other blocks.

**Comment by mojavelinux** [comment]
Here's some more background on how we arrived at this approach:

First, we reviewed the existing logic to confirm what margins were being added or collapsed. As expected, the logic is very inconsistent and, in many cases, uses guesswork patches to achieve the right spacing.

Ordered and unordered lists, especially when getting into nested lists, is particularly complicated. For example, a list item with primary content only (not multi-paragraph or block content) uses different spacing than a list item with paragraphs or blocks. This spacing complexity expands when one considers the content combinations of (great-great, great-, grand)parent, sibling, and child list items, and whether a list item is the last item in a list. As a person viewing a list, it's easy to identify the last list item in a list (i.e., the list terminus) visually. However, programmatically, the converter has to work through the AST until it identifies the greatest parent of an item and then that there are no other parents, siblings, or children list items following what may be the last list item. Description lists also have unique spacing requirements that include nested items and the possibility of terms without descriptions.

Then, we worked through what operations the converter would do to determine:

* What object owns a margin, assuming we apply bottom margins only?
* Which bottom margins should get collapsed depending on the next enclosed block?

For example, when a list item has complex content and is the last item in a list and the next object is a section heading is:

* the bottom margin of the last block that is part of the list item collapsed?
* the list item's bottom margin collapsed?
* the list's bottom margin collapsed?
* none of the bottom margins collapsed?

We then did the same exercise using top margins only. Both exercises included discussing what the possible results and problems would be if, in the future, we added keys for controlling some of the margins.

After working through a number of scenarios, we came up with the following plan:

. Implement smart bottom margins
.. Add new unit tests and update existing tests; our goal is to match the current spacing as closely as possible so the disruption to existing themes is minimal (but there will be visual changes when people switch to 2.0)
.. Removing bottom padding hacks in default theme; padding can now be applied consistently around the content
.. Remove top margins (since they don't take into account the context and thus behave inconsistently)
. Implement more consistent list item spacing for ordered, unordered, and description lists
.. Determine if compute+compare or recursion is more suitable for determining list terminus
.. Ensure that any application of a deferred bottom margin (i.e., dlist terms that don't have a description but are followed by another term) doesn't break the arrange_block logic.
. In a future issue, implement smart top margins using a parallel strategy
